### PR TITLE
Add protobuf-compiler and go generator

### PR DESCRIPTION
**Pull request by mangelajo**
This is beneficial to remove 1m30s extra added by the following PR

https://github.com/submariner-io/submariner/issues/1174

Signed-off-by: Miguel Angel Ajo <majopela@redhat.com>

### CODE REVIEW

**Comment by skitt** [comment]
We’ve typically refrained from adding project-specific tools to the base image; see the operator for example (although that needs reworking because it causes lots of lengthy image builds when doing local development).

**Comment by mangelajo** [comment]
> We’ve typically refrained from adding project-specific tools to the base image; see the operator for example (although that needs reworking because it causes lots of lengthy image builds when doing local development).

Yeah, I'm referencing that from my change in submariner. and adding a link from there to here:
```
diff --git a/Dockerfile.dapper b/Dockerfile.dapper
index d19263f..f910615 100644
--- a/Dockerfile.dapper
+++ b/Dockerfile.dapper
@@ -4,6 +4,10 @@ ENV DAPPER_ENV="REPO TAG QUAY_USERNAME QUAY_PASSWORD GITHUB_REF CLUSTERS_ARGS DE
     DAPPER_SOURCE=/go/src/github.com/submariner-io/submariner DAPPER_DOCKER_SOCKET=true
 ENV DAPPER_OUTPUT=${DAPPER_SOURCE}/output
 
+# NOTE: This is slow and can be removed by merging https://github.com/submariner-io/shipyard/pull/475
+RUN dnf install -y --nodocs --setopt=install_weak_deps=False golang-google-protobuf protobuf-compiler && dnf -y clean all
+
+
 WORKDIR ${DAPPER_SOURCE}

```
I noticed it's a little bit slow because it needs to download all the repo indexes.

**Comment by skitt** [comment]
@mkolesnik what do you think? IIRC it was you who objected to adding stuff for the operator.

**Comment by mangelajo** [comment]
> @mkolesnik what do you think? IIRC it was you who objected to adding stuff for the operator.

I'm fine either way, just wanted to bring it up because it bumps the submariner build by.. <measuring time, brb.. >

File: package/Dockerfile.shipyard-dapper-base [hunk 1]
```diff
@@ -38,15 +39,15 @@ ENV HOST_ARCH=${DAPPER_HOST_ARCH} ARCH=${DAPPER_HOST_ARCH} PATH=/go/bin:/usr/loc
...
+                   qemu-user-static python3-pip golang-google-protobuf protobuf-compiler && \
     npm install -g markdownlint-cli && \
     pip install j2cli[yaml] --user && \
     rpm -e --nodeps containerd npm python3-pip && \
     rpm -qa "selinux*" | xargs -r rpm -e --nodeps && \
     dnf -y clean all && \
     rm -f /usr/bin/{dockerd,lto-dump} \
           /usr/libexec/gcc/x86_64-redhat-linux/10/lto1 && \
-    find /usr/bin /usr/lib/golang /usr/libexec -type f -executable -newercc /proc -size +1M ! -name hyperkube | xargs upx ${UPX_LEVEL} && \
+    find /usr/bin /usr/lib/golang /usr/libexec -type f -executable -newercc /proc -size +1M ! -name hyperkube | grep -v protoc-gen-go | xargs upx ${UPX_LEVEL} && \
```
Comment (skitt) [inline package/Dockerfile.shipyard-dapper-base]:
```suggestion
    find /usr/bin /usr/lib/golang /usr/libexec -type f -executable -newercc /proc -size +1M ! -name hyperkube ! -name protoc-gen-go | xargs upx ${UPX_LEVEL} && \
```

**Review by skitt — CHANGES_REQUESTED** [review]
There’s already a pattern for filtering binaries ;-).

File: package/Dockerfile.shipyard-dapper-base (hunk 1 above)
Comment (mkolesnik) [inline package/Dockerfile.shipyard-dapper-base]:
```suggestion
    find /usr/bin /usr/lib/golang /usr/libexec -type f -executable -newercc /proc -size +1M ! -name hyperkube ! -name protoc-gen-go | xargs upx ${UPX_LEVEL} && \
```

Comment (mangelajo) [reply package/Dockerfile.shipyard-dapper-base]: oops I missed that :)

**Review by mkolesnik — CHANGES_REQUESTED** [review]

**Comment by mangelajo** [comment]
So, the extra time of not merging this is 1min 30 secs in submariner, not sure why it's so slow to download all the indexes TBH I'm on a 600Mbps connection may be it's the mirrors

**Comment by mkolesnik** [comment]
> So, the extra time of not merging this is 1min 30 secs in submariner, not sure why it's so slow to download all the indexes TBH I'm on a 600Mbps connection may be it's the mirrors

A possible mitigation is to have project specific base images that are cached in quay, e.g. `submariner-dapper-base` and `submariner-operator-dapper-base` or whatever names make sense.
Then our infra *should* take care of layer caching (but of course its a bit of a hassle when it doesnt work properly).

WFYT?

**Comment by skitt** [comment]
> > So, the extra time of not merging this is 1min 30 secs in submariner, not sure why it's so slow to download all the indexes TBH I'm on a 600Mbps connection may be it's the mirrors
> 
> A possible mitigation is to have project specific base images that are cached in quay, e.g. `submariner-dapper-base` and `submariner-operator-dapper-base` or whatever names make sense.
> Then our infra _should_ take care of layer caching (but of course its a bit of a hassle when it doesnt work properly).
> 
> WFYT?

Works for me, I have a PR-in-progress along those lines for the operator in fact ;-).

**Comment by mkolesnik** [comment]
> > A possible mitigation is to have project specific base images that are cached in quay, e.g. `submariner-dapper-base` and `submariner-operator-dapper-base` or whatever names make sense.
> > Then our infra _should_ take care of layer caching (but of course its a bit of a hassle when it doesnt work properly).
> > WFYT?
> 
> Works for me, I have a PR-in-progress along those lines for the operator in fact ;-).

@mangelajo can we make this a reality so that only consuming projects have their sh*t instead of sticking it all in the shipyard base image? :)

**Comment by mangelajo** [comment]
> > > A possible mitigation is to have project specific base images that are cached in quay, e.g. `submariner-dapper-base` and `submariner-operator-dapper-base` or whatever names make sense.
> > > Then our infra _should_ take care of layer caching (but of course its a bit of a hassle when it doesnt work properly).
> > > WFYT?
> > 
> > 
> > Works for me, I have a PR-in-progress along those lines for the operator in fact ;-).
> 
> @mangelajo can we make this a reality so that only consuming projects have their sh*t instead of sticking it all in the shipyard base image? :)

I think we'd be deviating from our full support goal tbh ;) 

I prefer to live with the extra 1m30s ... or the extra 3.5MB but I believe we shouldn't throw more CI work at ourselves for the time being, may be after 0.9 ? :)

**Comment by mkolesnik** [comment]
> So, the extra time of not merging this is 1min 30 secs in submariner, not sure why it's so slow to download all the indexes TBH I'm on a 600Mbps connection may be it's the mirrors

Them indexes are over 200MB so even compressed theyd add a significant weight on the base image, unfortunately

**Comment by skitt** [comment]
> > So, the extra time of not merging this is 1min 30 secs in submariner, not sure why it's so slow to download all the indexes TBH I'm on a 600Mbps connection may be it's the mirrors
> 
> Them indexes are over 200MB so even compressed theyd add a significant weight on the base image, unfortunately

Yup, that’s what I found out trying to split the image builds in Submariner; and those 200MiB can’t be recovered in subsequent layers.

**Comment by mangelajo** [comment]
As an alternative I could be downloading the binary .zip files of protoc for the submariner image too.

**Comment by mangelajo** [comment]
@mkolesnik @skitt what do you think about planning the base-image split for 0.10? 

I was looking at our schedule for 0.9 https://github.com/submariner-io/releases/issues/86 and well.. we are starting M2 at the release date given all the initial slow-down we had due to the 0.8.1 handling & the re-planning for "full support"

I can refactor that submariner repo dnf install into a curl & unzip for now :)

**Comment by mkolesnik** [comment]
Sure, but lets try not to accumulate too much technical debt

**Comment by mangelajo** [comment]
I will close this one for now, as I did the alternative of downloading on the submariner side with zips.

**Pull request closed without merging**